js/core/utils/shadow_dom.js

```diff
@@ -60,3 +60,55 @@ export function addShadowDomStyles($element) {
 
     root.adoptedStyleSheets = [constructedStyleSheet];
 }
+
+function isPositionInElementRectangle(element, x, y) {
+    const rect = element.getBoundingClientRect();
+
+    return rect && x >= rect.left && x < rect.right && y >= rect.top && y < rect.bottom;
+}
+
+function createQueue() {
+    let shiftIndex = 0;
+    const items = [];
+
+    return {
+        push(item) {
+            items.push(item);
+            return this;
+        },
+
+        shift() {
+            shiftIndex++;
+            return items[shiftIndex - 1];
+        },
+
+        get length() {
+            return items.length - shiftIndex;
+        }
+    };
+}
+
+export function getShadowElementsFromPoint(x, y, root) {
+    const result = [];
+    const elementQueue = createQueue().push(root);
+
+    while(elementQueue.length) {
+        const el = elementQueue.shift();
+
+        for(let i = 0; i < el.childNodes.length; i++) {
+            const childNode = el.childNodes[i];
+
+            // eslint-disable-next-line no-undef
+            if(childNode.nodeType === Node.ELEMENT_NODE &&
+               isPositionInElementRectangle(childNode, x, y) &&
+               // eslint-disable-next-line no-undef
+               getComputedStyle(childNode).pointerEvents !== 'none'
+            ) {
+                elementQueue.push(childNode);
+                result.push(childNode);
+            }
+        }
+    }
+
+    return result.reverse();
```

Comment:
seems like we don't need `result` array anymore - we can just get `items` array from our queue

Reply:
we can use items, but we still need to reverse them and exclude the shadow root element

Reply:
My idea is that we can not add an additional memory. After your last change we still have an additional array, but also make additional linear iteration.

Maybe like this?
`queue.getItems().reverse().slice(0, -1);